core/src/main/scala/matryoshka/patterns/Diff.scala

```diff
@@ -0,0 +1,73 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package matryoshka.patterns
+
+import scala.Unit
+
+import scalaz._, Scalaz._
+
+/** Represents diffs of recursive data structures.
+  */
+// more needs to be brought over from https://github.com/sellout/quasar/blob/3332dec0f6cdade26945b0db7215ce558a4f49cd/src/main/scala/slamdata/engine/analysis/diff.scala
+sealed trait Diff[T[_[_]], F[_], A]
+final case class Same[T[_[_]], F[_], A](ident: T[F]) extends Diff[T, F, A]
+final case class Similar[T[_[_]], F[_], A](ident: F[A]) extends Diff[T, F, A]
+final case class Different[T[_[_]], F[_], A](left: T[F], right: T[F])
+    extends Diff[T, F, A]
+final case class LocallyDifferent[T[_[_]], F[_], A](left: F[A], right: F[Unit])
+    extends Diff[T, F, A]
+final case class Inserted[T[_[_]], F[_], A](right: F[A])
+    extends Diff[T, F, A]
+final case class Deleted[T[_[_]], F[_], A](left: F[A])
+    extends Diff[T, F, A]
+final case class Added[T[_[_]], F[_], A](right: T[F])
+    extends Diff[T, F, A]
+final case class Removed[T[_[_]], F[_], A](left: T[F])
+    extends Diff[T, F, A]
+/** This node indicates that something about it is impossible to distinguish
+  * (e.g., a parameter with no Equal instance). It’s debatable whether something
+  * like this should be diffable, but I feel like diffing is generally more
+  * useful for debugging, and you let some things slide there, so …
+  */
+final case class Unknown[T[_[_]], F[_], A](left: F[A], right: F[Unit])
+    extends Diff[T, F, A]
+
+object Diff {
+  // TODO: implement low-prio Functor and Foldable with looser constraint on F
+  implicit def traverse[T[_[_]], F[_]: Traverse]: Traverse[Diff[T, F, ?]] =
+    new Traverse[Diff[T, F, ?]] {
+      def traverseImpl[G[_], A, B](
+        fa: Diff[T, F, A])(
+        f: A => G[B])(
+        implicit G: Applicative[G]):
+          G[Diff[T, F, B]] =
+        fa match {
+          case Same(ident) => G.point(Same(ident))
+          case Similar(ident) => ident.traverse(f) ∘ (Similar(_))
+          case Different(left, right) =>
+            G.point(Different(left, right))
+          case LocallyDifferent(left, right) =>
+            left.traverse(f) ∘ (LocallyDifferent(_, right))
+          case Inserted(right) => right.traverse(f) ∘ (Inserted(_))
+          case Deleted(left) => left.traverse(f) ∘ (Deleted(_))
+          case Added(right) => G.point(Added(right))
+          case Removed(left) => G.point(Removed(left))
+          case Unknown(left, right) =>
+            left.traverse(f) ∘ (LocallyDifferent(_, right))
+        }
+    }
+}
```

Comment:
I'm lost. This won't actually construct Diffs, right? I guess I have to map leaves for my type into `Diff` somehow? An example would really help to make this clear.